TESTS/host_tests/pyusb_basic.py

```diff
@@ -837,6 +930,544 @@ def control_stress_test(dev, log):
+                              .format(ENDPOINT_TYPE_NAMES[endpoint_type]))
+    raise_if_different(ep_out.wMaxPacketSize, ep_in.wMaxPacketSize, lineno(),
+                       'wMaxPacketSize not equal for OUT and IN {} endpoints.'
+                       .format(ENDPOINT_TYPE_NAMES[endpoint_type]))
+    return ep_out, ep_in
+
+
+def loopback_ep_test(ep_out, ep_in, payload_size):
+    """Send and receive random data using OUT/IN endpoint pair.
+
+    Verify that data received from IN endpoint is equal to
+    data sent to OUT endpoint.
+    Raise a RuntimeError if data does not match.
+    """
+    payload_out = array.array('B', (random.randint(0x00, 0xff) for _ in range(payload_size)))
+    ep_out.write(payload_out)
+    payload_in = ep_in.read(ep_in.wMaxPacketSize)
+    raise_if_different(payload_out, payload_in, lineno(), 'Payloads mismatch.')
+
+
+def random_size_loopback_ep_test(ep_out, ep_in, failure, error, seconds, log, min_payload_size=1):
+    """Repeat data transfer test for OUT/IN endpoint pair for a given time.
+
+    Set a failure Event if OUT/IN data verification fails.
+    Set an error Event if unexpected USB error occurs.
+    """
+    end_ts = time.time() + seconds
+    while time.time() < end_ts and not failure.is_set() and not error.is_set():
+        payload_size = random.randint(min_payload_size, ep_out.wMaxPacketSize)
+        try:
+            loopback_ep_test(ep_out, ep_in, payload_size)
+        except RuntimeError as err:
+            log(err)
+            failure.set()
+            return
+        except usb.USBError as err:
+            log(USB_ERROR_FMT.format(err, ep_out, ep_in, payload_size))
+            error.set()
+            return
+        time.sleep(0.01)
+
+
+def halt_ep_test(dev, ep_out, ep_in, ep_to_halt, log):
+    """OUT/IN endpoint halt test.
+
+    Verify that halting an endpoint at a random point of OUT or IN transfer
+    raises a USBError.
+    Raise a RuntimeError if halt fails or any unexpected error occurs.
+    """
+    MIN_HALT_DELAY = 0.01
+    MAX_HALT_DELAY = 0.1
+    POST_HALT_DELAY = 0.1
+    ctrl_error = Event()
+
+    for ep in (ep_out, ep_in):
+        try:
+            if (usb.control.get_status(dev, ep) == 1):
+                raise_unconditionally(lineno(), 'Endpoints must NOT be halted at the start of this test')
+        except usb.core.USBError as err:
+            raise_unconditionally(lineno(), 'Unable to get endpoint status ({!r}).'.format(err))
+
+    def timer_handler():
+        """Halt an endpoint using a USB control request."""
+        try:
+            usb.control.set_feature(dev, FEATURE_ENDPOINT_HALT, ep_to_halt)
+            if (usb.control.get_status(dev, ep_to_halt) != 1):
+                raise RuntimeError('Invalid endpoint status after halt operation')
+        except Exception as err:
+            ctrl_error.set()
+            log('Endpoint {:#04x} halt failed ({}).'.format(ep_to_halt.bEndpointAddress, err))
```

Comment:
I'm seeing this error intermittently in the test logs, but the tests are still passing:
```
Endpoint 0x01 halt failed (Invalid endpoint status after halt operation).
```

It appears that sometimes the next test starts running and unhalts the endpoint before `timer_handler` calls get_status. If I wait for the timer to finish before returning this problem does not occur.

Reply:
Good catch! I haven't spotted that. It appears I was naive enough to assume Python would kill a `Timer` thread created locally, when returning from a function.

Reply:
Fixed https://github.com/ARMmbed/mbed-os/blob/1de5b2c3c6191fbb58ee40a574c2da62d88c5da0/TESTS/host_tests/pyusb_basic.py#L1050